### PR TITLE
CASSANDRA-19336 5.0: Ensure that repair doesn't exceed repair_session_space by running repair jobs sequentially

### PR DESCRIPTION
**Pull request by adelapena**

### CODE REVIEW

File: src/java/org/apache/cassandra/repair/Scheduler.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                       ", durations=" + durations.longStream().average().orElse(-1) +
+                       ", score=" + score() +
+                       ", inflight=" + inflight +
+                       ", completed=" + completed +
+                       '}';
+            }
+        }
+    }
+
+    abstract class Task<T> extends AsyncFuture<T> implements Runnable
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/repair/Scheduler.java]: isn't this just `org.apache.cassandra.concurrent.FutureTask`?

File: src/java/org/apache/cassandra/repair/RepairJob.java [hunk 2]
```diff
@@ -189,30 +188,33 @@ public void onFailure(Throwable t)
...
+            {
+                createSyncTasks(paxosRepair, allSnapshotTasks, allEndpoints).addCallback((s, f) -> {
+                    if (f != null)
+                        tryFailure(f);
+                    else
+                        trySuccess(s);
+                });
+            }
+        };
+        session.validationScheduler.schedule(session.getId(), taskExecutor, syncTasks);
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/repair/RepairJob.java]: feel its cleaner to return `Future<T>` from `schedule`

File: src/java/org/apache/cassandra/repair/Scheduler.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        private final Map<TimeUUID, Group> groups = new HashMap<>();
+
+        LimitedConcurrentScheduler(int concurrentValidations, SharedContext ctx)
+        {
+            this.concurrentValidations = concurrentValidations;
+            this.ctx = ctx;
+        }
+
+        @Override
+        public synchronized void schedule(TimeUUID sessionId, Executor taskExecutor, Task<?> tasks)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/repair/Scheduler.java]: the usage is a single task, so all sessions will have the same `score`... so a simpler data structure could be used... this class makes the most sense when we have way to compare tasks

**Review by dcapwell — APPROVED** [review]

File: src/java/org/apache/cassandra/repair/Scheduler.java (hunk 1 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/repair/Scheduler.java]: This is slightly different due to how the `run` method sets a callback. I have moved the setting of that callback into the `Task` class to encapsulate it better.

File: src/java/org/apache/cassandra/repair/RepairJob.java (hunk 2 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/repair/RepairJob.java]: Done.

File: src/java/org/apache/cassandra/repair/Scheduler.java (hunk 3 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/repair/Scheduler.java]: I have changed it to a single queue of task-executor pairs.

**Pull request closed without merging**